### DISCUSSION TITLE
Why not just use Django + Alpine

### DISCUSSION THREAD

**Discussion by Vad1mo** [tetra-framework/tetra#8] [Q&A] [answered]
Hi, 

I used Vue, Alpine and Python (with Flask) in minor projects, not extensively though. I don't know Django that much (yet). 
This is not intended as a critique to your framework or its purpose, but your answer would help me understand and put the framework into context.

- Why not just use plain Django + Alpine
- Why not use Django + HTMX + Alpine or Hotwired + Alpine or Hotwired + Stimulus

thanks,

**Comment by samwillis** [comment, marked as answer]
Thanks for taking a look!

Tetra is doing four things:

- It proves the "glue" between Django and Alpine by making "public" methods (and attributes) that can be called directly from your alpine code. You don't need to build an API or have views that output HTML fragments for patching (as you do with vanilla HTMX). It should result in writing significantly less code.

- Also Tetra components are "statefull", further calls to public methods (on the backend) from the front end maintain the full state of the Python Component instance. This is like Livewire for PHP, although I believe (I'm not a PHP dev) we are able to support a wider range of types as Tetra uses Python Pickle for serialising. (the state is signed and encrypted to ensure it can't be tampered with or even read)
 
- It also encapsulates all your Python, HTML, CSS and JavaScript related to a component in the same file. This is inspired by front end toolkits like Vue.js with it's "single file components". I'm not aware of any frameworks that have taken this approach and included the backend, it's the driving force behind the concept. I'm calling it "close proximity of related concerns" and its similar (although not exactly the same as) "locality of behaviour" (https://htmx.org/essays/locality-of-behaviour/).

- It provides "modern" JavaScript build tooling, your components can `import` any JavaScript and it's all nicely built and packaged for you using esbuild (even during development when using `runserver`). The intention is to make it easy to use modern JS libraries without having to use webpack or similar.

Hope that helps, happy to answer any other questions.

**Reply by Vad1mo** [reply]
Thank you for your answer, so you are designing tetra to not need something like Hotwired or HTMX? because tetra provides this functionality?

**Reply by samwillis** [reply]
Exactly, not need for Hotwired or HTMX, it's a "batteries included" approach.

The one thing currently missing is a client side router, but this is planned, so where you might use Turbolinks to make a server rendered site feel like a SPA, Tetra will have that.

**Reply by khanakia** [reply]
@samwillis i was just wondering how do you store the Resumable state on server side ?

**Reply by samwillis** [reply]
It's not stored on the server, it's "Pickled", encrypted and sent to the browser, which then posts it back to the server on further requests. See: https://www.tetraframework.com/docs/state-security
